File: PA52.java
//Leetcode 143. Reorder List
// Time complexity : O(n) // O(n+n+n) linear time for each finding mid , reversing 2nd half of the linked list and 				merging both the parts of linked list.
//Space Complexity: O(1) // No extra space allocated.


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        //Find mid
        if(head==null || head.next==null){
            return ;
        }
        ListNode slow= head;
        ListNode fast= head;
        while(fast.next!=null && fast.next.next!=null){
            slow=slow.next;
            fast= fast.next.next;
        }
        
        
        //revesre the 2nd half
      
        ListNode lnode= reverse(slow.next);
        slow.next=null;

        slow= head;
        // merge both 
        
        while(lnode!=null){
            ListNode temp =  slow.next;
            slow.next= lnode;
            ListNode temp1=lnode.next;
            lnode.next= temp;
            slow=temp;
            lnode=temp1;
        }
        return ;
        
        
    }
    
    public ListNode reverse(ListNode root){
        ListNode prev= null;
        ListNode curr= root;
        if(root==null){
            return null;
        }
        ListNode fast = root.next;
        
        while(fast!=null){
            curr.next = prev;
        
            prev= curr;
            curr= fast;
            fast= fast.next;
       
            
        }
 
        curr.next = prev;
        return curr;
        
    }
}

File: PA_52.py
#LC 143 - Reorder List
#Time Complexity - O(n)
#Space Complxity - O(1)
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: None Do not return anything, modify head in-place instead.
        """
        if not head:
            return head
        def getMerged(head1,head2):
            temp = head1.next
            ans = head1
            while head2.next:
                head1.next = head2
                head2 = head2.next
                head1.next.next = temp
                head1 = temp
                temp = temp.next
            return ans 
        def getReversed(head):
            prev = None
            fast = head.next
            while fast:
                head.next = prev
                prev = head
                head = fast
                fast = fast.next
            head.next = prev
            return head
        fast = head
        slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        tmp = getReversed(slow)
        return getMerged(head,tmp)